### ISSUE TITLE
Rule Change: no-constructor-return to accept `return this;`

### ISSUE THREAD

**Issue by dolanmiu** [eslint/eslint#16727]
### What rule do you want to change?

[no-constructor-return](https://eslint.org/docs/latest/rules/no-constructor-return)

### What change to do you want to make?

Generate fewer warnings

### How do you think the change should be implemented?

A new option:

```
"no-constructor-return": [
    "error",
    {
        allowReturnThis: true,
    },
],
```

### Example code

```js
class B {
    constructor(f) {
        if (!f) {
            return this;
        }
    }
}
```

### What does the rule currently do for this code?

Currently, it throws an error when returning `this` in a constructor

### What will the rule do after it's changed?

It will give the option to allow `return this;` because it is valid JS.

Source:

https://stackoverflow.com/questions/3350215/what-is-returned-from-a-constructor#:~:text=The%20constructor%20returns%20the%20this%20object.

### Participation

- [X] I am willing to submit a pull request to implement this change.

**Comment by mdjermanovic** [comment]
Hi @dolanmiu, thanks for the issue!

 `return this;` in a constructor is the same as `return;`. This rule allows `return;` for flow control, but even though they're the same, I believe it's very uncommon to write `return this;` and that it looks like a mistake [this rule](https://eslint.org/docs/latest/rules/no-constructor-return) aims to warn about:

> In JavaScript, returning a value in the constructor of a class may be a mistake. Forbidding this pattern prevents mistakes resulting from unfamiliarity with the language or a copy-paste error.

Therefore, I think this rule shouldn't allow `return this;`. Curious what other team members think.

**Comment by nzakas** [comment]
I agree with @mdjermanovic. `return this` is almost always an indication that the author is unfamiliar with JavaScript and doesn't realize that `return this` is the default when nothing else is specified. 

And keep in mind that there is a lot of valid JavaScript that isn't good or proper to use. Most of our rules warn about valid JavaScript that is confusing.

One last note: If this pattern is important for you to allow, you can use a disable comment to opt-out of this warning, or else you can create your own version of the rule that doesn't exactly what you want.

**Comment by dolanmiu** [comment]
> Hi @dolanmiu, thanks for the issue!
> 
> `return this;` in a constructor is the same as `return;`. This rule allows `return;` for flow control, but even though they're the same, I believe it's very uncommon to write `return this;` and that it looks like a mistake [this rule](https://eslint.org/docs/latest/rules/no-constructor-return) aims to warn about:
> 
> > In JavaScript, returning a value in the constructor of a class may be a mistake. Forbidding this pattern prevents mistakes resulting from unfamiliarity with the language or a copy-paste error.
> 
> Therefore, I think this rule shouldn't allow `return this;`. Curious what other team members think.

It's not the same, please refer to:

https://github.com/dolanmiu/docx/issues/1858

`return;` returns undefined

**Comment by dolanmiu** [comment]
The use case is with flow control:

https://github.com/dolanmiu/docx/blob/master/src/file/paragraph/paragraph.ts#L52

Can this be re-openned?

**Comment by mdjermanovic** [comment]
Can you provide a JavaScript example where the difference between `return;` and `return this;` is observable when the code is run? Here's an example where the behavior is same:

```js
class A {
    constructor() {
        this.x = 1;
    }
}

class B extends A {
    constructor() {
        super();
        if (true) {
            this.y = 2;
            return; // same with `return this;`
        }        
    }
};

console.log(new B()); // B { x: 1, y: 2 }
```



If you're getting an unexpected behavior when your build output is executed, it could be because of an incorrect compiling/bundling/minifying/other processing of the source code.

**Comment by dolanmiu** [comment]
It is an issue with React production builds, could be their bundling

But even so, this is a valid use case, no? Or would this be out of scope?
